pkg/image/controller/controller.go

```diff
@@ -39,35 +33,105 @@ const retryCount = 2
 func (c *ImportController) Next(stream *api.ImageStream) error {
 	if !needsImport(stream) {
 		return nil
 	}
+
 	name := stream.Spec.DockerImageRepository
+	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
+
+	var tags map[string]string
+	// read default image metadata
+	if len(name) != 0 {
+		var (
+			err   error
+			retry bool
+		)
+		tags, err, retry = c.getDefaultTags(name, insecure)
+		if err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	}
+	if len(stream.Spec.Tags) == 0 {
+		// no tags specified, import all default
+		if err, retry := c.importDefaultTags(stream, tags, insecure); err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	} else {
+		// import only specified tags
+		limitedTags := make(map[string]string)
+		for tagName, specTag := range stream.Spec.Tags {
+			if tag, ok := tags[tagName]; ok && specTag.From == nil {
+				limitedTags[tagName] = tag
+			}
+		}
+
+		if err, retry := c.importDefaultTags(stream, limitedTags, insecure); err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+		if err, retry := c.importExternalTags(stream, insecure); err != nil {
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	}
+
+	// we've completed our updates
+	return c.done(stream, "", retryCount)
+}
+
+// getDefaultTags reads all tags from the stream's default registry,
+// returns tags, an error and bool flag specifying if the error is retryable
+func (c *ImportController) getDefaultTags(name string, insecure bool) (map[string]string, error, bool) {
+	var tags map[string]string
 
 	ref, err := api.ParseDockerImageReference(name)
 	if err != nil {
 		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
-		util.HandleError(err)
-		return c.done(stream, err.Error(), retryCount)
+		return tags, err, true
 	}
 
-	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
-
 	client := c.client
 	if client == nil {
 		client = dockerregistry.NewClient()
 	}
 	conn, err := client.Connect(ref.Registry, insecure)
 	if err != nil {
-		return err
+		return tags, err, false
 	}
-	tags, err := conn.ImageTags(ref.Namespace, ref.Name)
-	switch {
-	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-		return c.done(stream, err.Error(), retryCount)
-	case err != nil:
-		return err
+	tags, err = conn.ImageTags(ref.Namespace, ref.Name)
+
+	return tags, err, dockerregistry.IsRepositoryNotFound(err) || dockerregistry.IsRegistryNotFound(err)
+}
+
+// importDefaultTags imports tags from the stream's default registry
+// returns an error and bool flag specifying if the error is retryable
```

Comment:
After googling a while I didn't find a reasonable result one vs the other. But our sources prefer the version with 'y' :wink: